### DISCUSSION TITLE
PCL(Point Cloud) export

### DISCUSSION THREAD

**Discussion by nooskuyh** [NVlabs/instant-ngp#702] [Q&A]
![image](https://user-images.githubusercontent.com/47660152/174493273-54b103b3-5653-4819-a3b7-74ddc0514db9.png)

I want to export the point cloud shown in the testbed.
so i have tried this code,

```python
import argparse
import os
import commentjson as json

import numpy as np

import sys
import time

from common import *
from scenes import scenes_nerf, scenes_image, scenes_sdf, scenes_volume, setup_colored_sdf

from tqdm import tqdm

import pyngp as ngp # noqa
from plyfile import PlyData, PlyElement
from scenes import scenes_nerf, scenes_image, scenes_sdf, scenes_volume, setup_colored_sdf

mode = ngp.TestbedMode.Nerf
testbed = ngp.Testbed(mode)
testbed.load_snapshot("ring_snap.msgpack")
testbed.aabb = ngp.BoundingBox([0.320, 0.320, 0.320], [0.680, 0.680, 0.680])
def round_up_to_base(x, base=10):
    return x + (base - x) % base

def get_marching_cubes_res(res_1d: int, aabb:  ngp.BoundingBox ) -> np.ndarray:
	scale = res_1d / (aabb.max - aabb.min).max()
	res3d = (aabb.max - aabb.min) * scale + 0.5
	res3d = round_up_to_base(res3d.astype(np.int32), 16)
	return res3d

mc = testbed.compute_marching_cubes_mesh(resolution=get_marching_cubes_res(512, testbed.aabb), aabb=testbed.aabb, thresh=2)


vertex = np.array(list(zip(*mc["V"].T)), dtype=[('x', 'f4'), ('y', 'f4'), ('z', 'f4')])
vertex_color = np.array(list(zip(*((mc["C"] * 255).T))), dtype=[('red', 'u1'), ('green', 'u1'), ('blue', 'u1')])

n = len(vertex)
assert len(vertex_color) == n

vertex_all = np.empty(n, vertex.dtype.descr + vertex_color.dtype.descr)

for prop in vertex.dtype.names:
    vertex_all[prop] = vertex[prop]

for prop in vertex_color.dtype.names:
    vertex_all[prop] = vertex_color[prop]

ply = PlyData([PlyElement.describe(vertex_all, 'vertex')], text=False)

ply.write('nerf_pc.ply')
```
![image](https://user-images.githubusercontent.com/47660152/174493903-b3c1ed24-e1ec-4f5c-a8de-32507b931ca0.png)

but it's not work well, I just got Sparse Pointcloud...😭
Next, I tried to export the Mesh(obj) with this Python code not PCL.

and i want mesh object. so tried this code.
```python

python3 ./scripts/run.py --mode nerf --scene ./data/nerf/fox/ --save_snapshot saved/fox.msgpack --save_mesh saved/fox.ply --gui --train

```
However, the resulting obj file was different from the result exported from the gui.
I think the reason is that the aabb range cannot be set.

1. How can i export pointcloud seems like testbed?
2. How can i properly export a mesh in no gui?

finally, i want to show 3d object in web browser...

**Comment by kafei123456** [comment]
Excuse me, have you solved the problems?

**Comment by nooskuyh** [comment]
No, i can't.
it's impossible question.
because point clouds that the model makes is only specific presentation. (not include point of view that we can't see). so i recommend you to look up MobileNerf Model.
________________________________

**Reply by kafei123456** [reply]
Oh, But MobileNerf need to many GPU. I don't know how to train it using only one GPU.
